Handling columns in wasm module

alban opened the pull request:
# Handling columns in wasm module

Part of https://github.com/inspektor-gadget/inspektor-gadget/issues/2202

## How to use

```
$ go run -exec 'sudo -E' ./cmd/ig/... image build -t ghcr.io/inspektor-gadget/gadget/trace_dns:latest gadgets/trace_dns/
INFO[0000] Experimental features enabled                
Successfully built ghcr.io/inspektor-gadget/gadget/trace_dns:latest@sha256:4513df2dfd78bde41fee7f46dc35db2f0923d5a266002ef2ab920ca486e246d3

$ go run -exec 'sudo -E' ./cmd/ig/... run ghcr.io/inspektor-gadget/gadget/trace_dns:latest
INFO[0000] Experimental features enabled                
RUNTIME.CONTAINERNAME  SRC               DST               PID     TASK  NAME            QR  PKT_TYPE RCODE    QTYPE
gifted_torvalds        172.17.0.4:48815  192.168.0.1:53    165520  ping  wikipedia.org.  Q   OUTGOING NoError  A
gifted_torvalds        172.17.0.4:48815  192.168.0.1:53    165520  ping  wikipedia.org.  Q   OUTGOING NoError  AAAA
gifted_torvalds        192.168.0.1:53    172.17.0.4:48815  165520  ping  wikipedia.org.  R   HOST     NoError  A
gifted_torvalds        192.168.0.1:53    172.17.0.4:48815  165520  ping  wikipedia.org.  R   HOST     NoError  AAAA
```

## Testing done

[Describe the testing you have done before submitting this PR. Please include both the commands you issued as well as the output you got.]

mauriciovasquezbernal commented:
I'll create an independent PR for the "Experiment: Avoid decoding BTF on the client" commit so we can review it easier.

alban commented:
PR rebased on top of `main` (@mauriciovasquezbernal did all the work), now that #2269 is merged.

mauriciovasquezbernal commented on pkg/gadgets/run/tracer/wasm.go:
```diff
@@ -0,0 +1,62 @@
+	drop bool
+}
+
+func (t *Tracer) newWasmHost(l logger.Logger) func(ctx context.Context, binding, namespace, operation string, payload []byte) ([]byte, error) {
+	return func(ctx context.Context, binding, namespace, operation string, payload []byte) ([]byte, error) {
+		cookie := ctx.Value(wasmHostCallContextKey).(*wasmHostCallContext)
+		switch binding {
+		case "ig":
+			switch namespace {
+			case "event":
+				switch operation {
+				case "drop":
+					cookie.drop = true
+					return nil, nil
+				}
+			}
+		}
+		l.Warnf("HostCall for %s/%s/%s not implemented", binding, namespace, operation)
+
+		return nil, fmt.Errorf("HostCall for %s/%s/%s not implemented", binding, namespace, operation)
+	}
+}
```
Is this to allow the wasm program to drop an event? It's not used in any gadget yet, right?

alban replied:
Correct.

mauriciovasquezbernal commented on pkg/gadgets/run/tracer/run.go:
```diff
@@ -857,12 +863,14 @@ func calculateColumnsForClient(
 		}
 
 		columns = append(columns, col)
+
```
nit: remove (it was probably me during the rebasing)

mauriciovasquezbernal commented on pkg/gadgets/run/tracer/tracer.go:
```diff
@@ -360,8 +407,27 @@ func (t *Tracer) processEventFunc(gadgetCtx gadgets.GadgetContext) func(data []b
 		typ   endpointType
 	}
 
+	type wasmColumnDef struct {
+		setter func(*types.Event, string)
+		name   string
+		start  uint32
+		size   uint32
+	}
```
I'm wondering if we could use an approach similar to the enumeration support: some "setters" are created, then the tracer only calls them. I tried it on https://github.com/inspektor-gadget/inspektor-gadget/commit/58e5e7b123cac3c5e299573595b8a48f59a08140.

mauriciovasquezbernal reviewed (commented):
I did a first pass. It already looks very good. 

I think first and second commit should be squashed. (The second commit is undoing almost all changes on the first one)

Also, the last commit states:

> The following columns are not handled in the wasm module:
name, qr, qtype, rcode, pkt_type

I suppose it should be "now" instead of "now", right?

Finally, are you planning to add some documentation about it?

alban commented on gadgets/trace_dns/program.go:
```diff
@@ -1,9 +1,203 @@
+	return nil, nil
+}
+
+// pkt_type definitions:
+// https://github.com/torvalds/linux/blob/v5.14-rc7/include/uapi/linux/if_packet.h#L26
+var pktTypeNames = []string{
+	"HOST",
+	"BROADCAST",
+	"MULTICAST",
+	"OTHERHOST",
+	"OUTGOING",
+	"LOOPBACK",
+	"USER",
+	"KERNEL",
+}
```
It requires a fix that I filed in a separate PR: https://github.com/inspektor-gadget/inspektor-gadget/pull/2298

alban replied:
pktTypeNames and rcodes are now done as enums.

qtype remains in Go/Wasm because enums can't have `*` characters.

alban commented on gadgets/trace_dns/program.go:
```diff
@@ -1,9 +1,203 @@
+}
+
+func Init(payload []byte) ([]byte, error) {
+	//wapc.ConsoleLog("Hello from Wasm!")
```
I removed the comment but I kept the empty function: it seems that the host is not able to inspect the wasm module to know if the function exists at all. If it does not exist, it returns an error, but unless we do string parsing, we don't know if Init() returned an error or if it does not exist.

So I am suggesting to make the Init() function mandatory. The difference with the `main()` function is that it can receives custom parameters, which can be useful for some gadgets later on.

mauriciovasquezbernal commented on pkg/oci/oci.go:
```diff
@@ -108,23 +109,29 @@ func GetGadgetImage(ctx context.Context, image string, authOpts *AuthOptions, pu
 		return nil, fmt.Errorf("getting ebpf program: %w", err)
 	}
 
+	wasm, err := getLayerFromManifest(ctx, imageStore, manifest, wasmObjectMediaType)
+	if err != nil {
+		return nil, fmt.Errorf("getting wasm program: %w", err)
+	}
+
```
This layer is optional, it shouldn't fail if the layer is not found.

alban replied:
I'll address this in https://github.com/inspektor-gadget/inspektor-gadget/pull/2355#discussion_r1448655497 first.

mauriciovasquezbernal commented on pkg/gadgets/run/tracer/tracer.go:
Would you mind creating an independent PR for this one to fix the issue ASAP?

alban replied:
The commit about enum fix is already in main via https://github.com/inspektor-gadget/inspektor-gadget/pull/2298

mauriciovasquezbernal replied:
I went too fast without checking. Thanks for the quick reply.

alban commented:
I rebased, fixed the conflicts and addressed some of the reviews.

TODO:
- [ ] Documentation
- [ ] Avoid handling individual columns in individual Wasm functions but handle full events instead
